### ISSUE TITLE
(Soft-)deprecate `init` for `sum!` etc?

### ISSUE THREAD

**Issue by tkf** [JuliaLang/julia#36266]
As @timholy pointed out in #35839, the meaning of `init` in in-place multi-dimensional reducers like `sum!` and `count!` are incompatible with `reduce` etc.

Now that `sum(...; init)` #36188 is merged and would be shipped with Julia 1.6, it's a bit concerning that `init` in `sum!(r, A; init = false)` and `sum(A; init = false)` means something completely different.  I think it might make sense to deprecate or "soft-deprecate" `init` for `sum!` etc.  By "soft-deprecate" I mean to add a new keyword argument that has `init` as the fallback without `depwarn`.  This new keyword argument would be documented as a preferred way to control the initialization of the destination array.  Since `sum!` etc. do not actually document `init`, going directly to deprecation might be OK.  But, since this keyword argument exists for a long time, I'm inclined to treat this as a fully-fledged public API.  (This is a bit tricky case because it's unclear if performance-breaking but semantics-preserving change should be treated as non-breaking.)

For the actual name of the new keyword argument, how about `sum!(r, A; fill = true)` instead of `sum!(r, A; init = true)`?

**Comment by tkf** [comment]
Looking at it again, now I'm not so sure about `fill` as the new keyword argument. Brainstorming some other names:

```
fillinit
initialize
initfirst
needinit
```

Alternatively, it also makes sense to rename `init` of `reduce` etc. to `identity`, `id`, `neutral`, or something.  We need something similar to `init` for `foldl`/`foldr`/`accumulate`, though.

(cc'ing other people who might be interested in this: @nalimilan @mbauman @JeffBezanson)

**Comment by mcabbott** [comment]
How about `sum!(r, A; keep = false)` or `sum!(r, A; discard = true)`?

**Comment by tkf** [comment]
I think `discard` is a good candidate! I guess `keep` is kind of a same idea but `keep`ing some value happens only if the corresponding region of the input array is zero (or `NaN`/`missing` is in the destination). I think `discard` is better because `discard`ing old values happen always.

**Comment by timholy** [comment]
I also like `discard`, great suggestion @mcabbott and thanks @tkf for spearheading this and so many other advances!

**Comment by StefanKarpinski** [comment]
Doesn't the keyword argument to `sum!` control whether the array gets filled first or not? This keyword isn't documented anywhere that I can see and even figuring out what it does requires perusing some fairly confusing code in `reducedim.jl`

**Comment by tkf** [comment]
Right, `init` for `sum!` is not used.  So very strictly speaking it's not a public API.  But I don't know how it's used in the wild.

In Julia 1.5, we'll have `count!` which mentions `init`:

https://github.com/JuliaLang/julia/blob/13b07fcc57d268b37234d8b8753b4ef3447be8db/base/reducedim.jl#L393-L400

> I don't get how `keep` or `discard` are more appropriate than `fill`.

Well, I'm OK with `fill`, too. I thought `discard` was better because `fill` _could_ mean filling something else (e.g., filling `missing`/`NaN`; although I think the chance for such misunderstanding is very low). OK, maybe it can be said for `discard` ("discarding" `missing`s). Naming is damn hard....

I actually think `initialize` is slightly better than `discard`.  I think it also avoid confusion with filling `missing`/`NaN`.

**Comment by timholy** [comment]
Another option is `reset`

**Comment by mbauman** [comment]
I've never used a bang reduction with `init`, and — even knowing that this was a boolean flag based on the first post — my first instinct was that this flag would work oppositely as it does (that is, if true, use the existing values in the output array as the reducer's `init` kwarg).  So that's another motivation to change this.

In some senses, I think a pretty good API would be one in which we defaulted to using the values in the output array and if an `init` kwarg is provided we `fill!(output, init)` before performing the reduction.  Unfortunately that's definitely and hugely breaking.

Alternatively, we could use a sentinel like `nothing` or `undef` to opt out of initialization and allow `init`'ing arbitrary values.  Again, though, that's changing the meaning of `init` in the wild.  It's not [used much though](https://juliahub.com/ui/Code?q=%5C%21%5C%28.%2a%5Cbinit%5Cs%2a%3D&r=true) — that's a fairly conservative search and really only flags GPUArrays (but it misses multi-line arg lists).

If we're going to use a boolean flag, `reset` is the best option yet. I wish we could also convey the zeroing property, but neither `zero` nor `zero!` make for good kwargs.  We could alternatively define this as a function that "preprocesses" the output array, with a default of `out->fill!(out, zero(eltype(out)))`... but specifying `preprocess=identity` doesn't feel like a good way of opting out of this behavior.

**Comment by StefanKarpinski** [comment]
In hindsight, it seems like it would be best if the out-of-place accumulators were defined to always increment the target array: if you want to start at zero, pass in a zeroed array.

**Comment by tkf** [comment]
`maximum!` does something a bit different than filling the identity element (it uses the first slice of the input array).  I think auto-initialization is a useful API as initial value handling can be a headache sometimes.

**Comment by mbauman** [comment]
> `maximum!` does something a bit different than filling the identity element (it uses the first slice of the input array)

You could still think of this as filling an identity element — the "maximative" identity is the `typemin`, yeah?

**Comment by tkf** [comment]
Well, I'm a bit ambivalent about using `typemin`/`typemax` for `maximums`/`minimum`.  I think erroring out is a bit better.  This is because the "semantic" domain of the function may not be captured by the type. For example, if you have

```julia
ys .= sin.(xs)
maximum(ys)
```

it's in some sense not desirable to get `-Inf` when `xs` is empty.   It'd be nice if it's -1 but of course, that's rather impossible.

**Comment by nalimilan** [comment]
> In Julia 1.5, we'll have `count!` which mentions `init`:

Should we remove the mention of that keyword argument before people start using it?

**Comment by tkf** [comment]
Yeah, good point. Now that everyone agrees that rename is the way to go (and converging to `reset` as the new name?), it makes sense to at least hide it in the documentation.  I just opened #36305 that does that.

**Comment by tkf** [comment]
Is everyone OK with renaming `init` to `reset`?

**Comment by StefanKarpinski** [comment]
Can someone write a sentence explaining what the artist formerly known as `init` does?

**Comment by tkf** [comment]
`f!(r, A; init)` does the following before the reduction iff `init = true`:

* `sum!`, `count!`: `fill!(r, zero(eltype(r)))`
* `prod!`: `fill!(r, one(eltype(r)))`
* `all!`: `fill!(r, true)`
* `any!`: `fill!(r, false)`
* `maximum!`, `minimum!`: fill `r` with the first slice of `A`

**Comment by mbauman** [comment]
In the context of 0-element reductions:

* `init` is used as the output. If not provided, error.

In the context of 1+-element reductions:

* `init` is used as an argument to the first call to the reducer. If not provided and there's only one element, just return that one element itself.

In the context of bang (in-place) reductions, it's really hard.  Here's the best I can do:

* This flag chooses whether to use the existing values in the output as the reducer's `init` argument or not.

**Comment by StefanKarpinski** [comment]
The first two are the ones we're keeping the `init` term for (and it fits) so I'm not worried about them. It's the last tough one I'm trying to get a handle on and I fear that `reset` is no clearer, which is why I wanted to play the "write a sentence that explains what it does" game since there's usually a word in that sentence that is a good choice for the argument name. One thing that keeps occurring to me is `initialized` with the opposite meaning of the current name that indicates whether the incoming argument is already initialized or not (if it isn't, then the reducer should initialize it).

**Comment by tkf** [comment]
I created a WIP PR for this #36332.  I updated the docstring to what it does.  I also added a few doctest demos.  I guess it'd help the game?

Here is the canonical phrase I used in the PR:

> If `reset` is `true`, values in `r` are discarded.  Otherwise, they are used as the initial value of the accumulation.

(`reset` is just a placeholder for now)

**Comment by mbauman** [comment]
One option would be to use the word `init` but pass either _an array_ the same shape as `out` or `nothing`.  This would allow folks to pass `out` itself, and we could still easily deprecate `init::Bool`.

**Comment by tkf** [comment]
Hm... I'm a bit confused. Do you mean `sum!(A; out = r)` instead of `sum!(r, A; init = false)`?  What is the signature for `sum!(r, A)`?

**Comment by mbauman** [comment]
No, I mean you'd call `sum!(r, A; init=r)` in cases where you want to preserve the values in `r` as the `init` state of the reductions.  By default we'd pass `init=nothing`, which would be the sentinel that flags not having an `init` in each reduction.

**Comment by tkf** [comment]
If the `out` array is not the one mutated, isn't it incompatible with the name out(put)?  (Or at least that's [how it is used in Numpy](https://numpy.org/doc/stable/reference/generated/numpy.dot.html) so I was confused.)

**Comment by tkf** [comment]
@StefanKarpinski's `initialized` sounds good to me too, BTW.

**Comment by mbauman** [comment]
Hunh?  This is just determining what the output array gets pre-filled with.  E.g.,

```Julia
function sum!(r, A; init=zeros(size(r)))
    if init === r
        # do nothing
    elseif init isa Bool
        # depwarn
    else
        r .= init
    end
    # now proceed with reduction in-place using the values in `r` as the first argument
    return r
end
```

**Comment by tkf** [comment]
Oh, I misunderstood your proposal. I thought you were suggesting to use `out` as the name of the keyword argument.

**Comment by tkf** [comment]
Wondering about `init :: AbstractArray` API, I still think flag-based interface is easier to compose.  For example, I think

```julia
r = sum!(preprocess(), A; reset = false)
```

is much cleaner than

```julia
r = preprocess()
sum!(r, A; init = r)
```

**Comment by mbauman** [comment]
OK, so finding the language here becomes a bit easier if we concretely identify *how* the elements in `R` are actually used in the expression `sum!(R, A; _______)`:

* It's *not* as an `init`, despite what I [had written in 2020](https://github.com/JuliaLang/julia/issues/36266#issuecomment-644973124).  In fact, it can't be.  The scalar `init` must be a *neutral* value that may be incorporated into the reduction one or more times.  That's neither useful nor pragmatic; you'd need to keep `R` unchanged throughout the duration of the reduction (or copy it) in order to use its existing values multiple times.  And if if it must be a neutral value, then that goes against basically every usage of it!
* It's *not* as though it were the first element in `A`.  Unlike the elements in `A`, we don't — and you wouldn't want to! — apply the transforming function to the values in `R`.  A concrete use case here is something like an iterative `minimum!(abs2, R, A; ______)`.  You don't want to square the values in `R` — you're likely taking the previous output of that expression (which was already squared) and _updating_ it.

So perhaps that's the flag name here. This is `sum!(R, A; update=true)`.  Or maybe some synonym of update.

Now an interesting question is if `sum!([0], []; update=true)` should be an error — currently it is not.  Pragmatically, I think it'll be a lot easier to implement if it is an error.  Interestingly, `init` becomes an orthogonal concept now; that could be permitted with `sum!([0], [], update=true, init=0)`.

**Comment by LilithHafner** [comment]
Triage likes `update` as an alternative name, if needed.
